# DSC 80: Lab 04

### Due Date: Tuesday February 4th, Midnight (11:59 PM)

## Instructions
Much like in DSC 10, this Jupyter Notebook contains the statements of the problems and provides code and markdown cells to display your answers to the problems. Unlike DSC 10, the notebook is *only* for displaying a readable version of your final answers. The coding work will be developed in an accompanying `lab*.py` file, that will be imported into the current notebook.

Labs and programming assignments will be graded in (at most) two ways:
1. The functions and classes in the accompanying python file will be tested (a la DSC 20),
2. The notebook will be graded (for graphs and free response questions).

**Do not change the function names in the `*.py` file**
- The functions in the `*.py` file are how your assignment is graded, and they are graded by their name. The dictionary at the end of the file (`GRADED FUNCTIONS`) contains the "grading list". The final function in the file allows your doctests to check that all the necessary functions exist.
- If you changed something you weren't supposed to, just use git to revert!

**Tips for working in the Notebook**:
- The notebooks serve to present you the questions and give you a place to present your results for later review.
- The notebook on *lab assignments* are not graded (only the `.py` file).
- Notebooks for PAs will serve as a final report for the assignment, and contain conclusions and answers to open ended questions that are graded.
- The notebook serves as a nice environment for 'pre-development' and experimentation before designing your function in your `.py` file.

**Tips for developing in the .py file**:
- Do not change the function names in the starter code; grading is done using these function names.
- Do not change the docstrings in the functions. These are there to tell you if your work is on the right track!
- You are encouraged to write your own additional functions to solve the lab! 
    - Developing in python usually consists of larger files, with many short functions.
    - You may write your other functions in an additional `.py` file that you import in `lab**.py` (much like we do in the notebook).
- Always document your code!

### Importing code from `lab**.py`

* We import our `.py` file that's contained in the same directory as this notebook.
* We use the `autoreload` notebook extension to make changes to our `lab**.py` file immediately available in our notebook. Without this extension, we would need to restart the notebook kernel to see any changes to `lab**.py` in the notebook.
    - `autoreload` is necessary because, upon import, `lab**.py` is compiled to bytecode (in the directory `__pycache__`). Subsequent imports of `lab**` merely import the existing compiled python.

In [2]:
%load_ext autoreload
%autoreload 2

In [3]:
import lab04 as lab

In [4]:
import os
import pandas as pd
import numpy as np

## Login time questions

Imagine that you own an online store and you'd like to monitor the visits to your site. You've collected some data that you store in login_table.csv. It contains the information about different login dates and times for different users. Some users are unique, some visited your store multiple times.

You need to answer a few questions below in order to understand the login patters of your users.

In [4]:
fp = os.path.join('data', 'login_table.csv')
login = pd.read_csv(fp)

**Question 1**

Write a function `latest_login` which takes in a dataframe like `login` and outputs a dataframe, indexed by `Login Id`, of the login that occurs at the latest time-of-day for each user. Latest time-of-day is as it says: the latest time, regardless of how recent the day is.

For example, if a user always logs in once per day at Noon, but her most recent log in happened to be at 8:00PM, then her latest log-in time becomes 8:00PM.

Note: you do not need to import datetime. Look into how pandas has its own built in methods for dates.

In [21]:
fp = os.path.join('data', 'login_table.csv')
login = pd.read_csv(fp)
#login.groupby('Login Id').max()

In [22]:
result = lab.latest_login(login)

In [23]:
len(result)

433

In [24]:
result.loc[381, "Time"].hour > 12

True

In [25]:
result.iloc[381]

Time    13:09:38
Name: 1229, dtype: object

**Question 2**

As a site owner, you would like to see how often users return to your site. You've noticed that there are users who have several logins and users who logged in only once. Of those users that logged in more than once, you are interested in finding the shortest amount of time elapsed between two consecutive logins for each of these users.

Write a function `smallest_ellapsed` which takes in a dataframe like `login` and outputs a dataframe, indexed by Login ID, containing the shortest time elapsed for each user. Any users who haven't logged in more than once should not be included in the output.

In [11]:
fp = os.path.join('data', 'login_table.csv')
login = pd.read_csv(fp)
login['Time'] = pd.to_datetime(login['Time'])
login
table = login.groupby('Login Id').apply(lambda x: x['Time'].sort_values(ascending=False).diff().min())
# table.apply(lambda x: x.diff()).dt.days

In [12]:
table

Login Id
381    -137 days +00:44:46
393    -148 days +01:08:32
412     -22 days +00:42:53
413     -23 days +19:21:26
419     -11 days +01:49:50
425    -165 days +23:30:07
427    -136 days +14:39:37
428                    NaT
442    -181 days +03:46:15
443     -74 days +00:17:10
447     -74 days +23:50:44
449     -49 days +03:44:14
450     -75 days +06:22:20
451     -14 days +23:35:22
452      -3 days +04:38:11
453    -109 days +02:43:56
457     -80 days +22:29:47
458     -20 days +21:26:45
462     -90 days +20:26:52
463     -32 days +02:57:32
464    -158 days +00:15:18
466     -49 days +07:11:48
468     -45 days +18:38:40
470    -176 days +23:59:33
472     -64 days +02:53:23
473      -1 days +23:52:26
474    -149 days +18:18:12
478     -30 days +00:11:43
482                    NaT
483                    NaT
               ...        
1263                   NaT
1264     -1 days +23:40:07
1265     -1 days +22:01:52
1266                   NaT
1269     -1 days +23:59:22
1270               

In [49]:
# result = lab.smallest_ellapsed(login)

In [28]:
18 < result.loc[1233, "Time"].days < 23

True

# Pivot tables

**Question 3**

The pivot table allows you to group the entries of a dataframe into a two-dimensional table that provides a (multidimensional) summarization of the data. You are given a simple dataset, `sales.csv`, and are asked to solve a few simple problems using a `pivot table`.  

We have provided the outline for your pivot tables. Your values will be different.

* Write a function `total_seller` that takes `sales` dataframe and returns a pivot table that contains a total for each seller, indexed by a name.

<img src="imgs/image_0.png" width="15%"/>

* Write a function `product_name` that takes `sales` dataframe and returns a pivot table that contains a total for each seller, indexed by a product.

<img src="imgs/image_1.png" width="25%"/>

* Write a function `count_product` that takes `sales` dataframe and returns a pivot table that contains the total amount of items sold product wise, name wise per date. Replaces `NaNs` with 0s. 

<img src="imgs/image_2.png" width="35%"/>

* Write a function `total_by_month` that takes `sales` dataframe and returns a pivot table that contains the total amount name wise, product wise per `month`. Replaces `NaNs` with 0s.

<img src="imgs/image_3.png" width="40%"/>


Note: Here is a <a href = "https://jakevdp.github.io/PythonDataScienceHandbook/03.09-pivot-tables.html"> link </a> to a great source that provides an overview of the pivot tables with mane examples from the Titanic dataset. 

If your "Total" title is on the right side of the DataFrame instead of the left as seen above, as long as you are passing the doctests you should be fine. This is likely due to an issue with pandas versions.

In [47]:
fp = os.path.join('data', 'sales.csv')
df = pd.read_csv(fp)
df

Name Product        Date  Total
0   Smith    book  01.01.2012    200
1   Jones     pen  02.20.2013    300
2   Trump   hotel  03.03.2015    100
3   Smith    book  05.10.2013   2000
4   Jones    book  02.20.2013     30
5   Trump    boat  03.30.2017    700
6   Smith   ruler  07.05.2014   2100
7   Jones   ruler  02.25.2015    350
8   Trump    book  03.03.2015   1000
9   Smith     pen  01.01.2012   2500
10  Jones   ruler  02.20.2013   3000
11  Trump     pen  03.03.2015    150

In [12]:
out = lab.total_seller(df)

In [79]:
out.index.dtype

dtype('O')

In [80]:
out["Total"].sum() < 15000

True

In [13]:
pd.pivot_table(df, values=['Total'], index=['Product','Name'],
                    columns=['Date'], aggfunc=np.sum)

Total                                              \
Date          01.01.2012 02.20.2013 02.25.2015 03.03.2015 03.30.2017   
Product Name                                                           
boat    Trump        NaN        NaN        NaN        NaN      700.0   
book    Jones        NaN       30.0        NaN        NaN        NaN   
        Smith      200.0        NaN        NaN        NaN        NaN   
        Trump        NaN        NaN        NaN     1000.0        NaN   
hotel   Trump        NaN        NaN        NaN      100.0        NaN   
pen     Jones        NaN      300.0        NaN        NaN        NaN   
        Smith     2500.0        NaN        NaN        NaN        NaN   
        Trump        NaN        NaN        NaN      150.0        NaN   
ruler   Jones        NaN     3000.0      350.0        NaN        NaN   
        Smith        NaN        NaN        NaN        NaN        NaN   

                                     
Date          05.10.2013 07.05.2014  
Product Name                         
boat    Trump        NaN        NaN  
book    Jones        NaN        NaN  
        Smith     2000.0        NaN  
        Trump        NaN        NaN  
hotel   Trump        NaN        NaN  
pen     Jones        NaN        NaN  
        Smith        NaN        NaN  
        Trump        NaN        NaN  
ruler   Jones        NaN        NaN  
        Smith        NaN     2100.0

In [85]:
out = lab.product_name(df)
out.size

15

In [86]:
out.loc["pen"].isnull().sum()

0

In [30]:
df.groupby(['Product', 'Name']).size()

Product  Name 
boat     Trump    1
book     Jones    1
         Smith    2
         Trump    1
hotel    Trump    1
pen      Jones    1
         Smith    1
         Trump    1
ruler    Jones    2
         Smith    1
dtype: int64

In [116]:
out = lab.count_product(df)
out

Date           01.01.2012  02.20.2013  02.25.2015  03.03.2015  03.30.2017  \
Product Name                                                                
boat    Trump         0.0         0.0         0.0         0.0         1.0   
book    Jones         0.0         1.0         0.0         0.0         0.0   
        Smith         1.0         0.0         0.0         0.0         0.0   
        Trump         0.0         0.0         0.0         1.0         0.0   
hotel   Trump         0.0         0.0         0.0         1.0         0.0   
pen     Jones         0.0         1.0         0.0         0.0         0.0   
        Smith         1.0         0.0         0.0         0.0         0.0   
        Trump         0.0         0.0         0.0         1.0         0.0   
ruler   Jones         0.0         1.0         1.0         0.0         0.0   
        Smith         0.0         0.0         0.0         0.0         0.0   

Date           05.10.2013  07.05.2014  
Product Name                           
boat    Trump         0.0         0.0  
book    Jones         0.0         0.0  
        Smith         1.0         0.0  
        Trump         0.0         0.0  
hotel   Trump         0.0         0.0  
pen     Jones         0.0         0.0  
        Smith         0.0         0.0  
        Trump         0.0         0.0  
ruler   Jones         0.0         0.0  
        Smith         0.0         1.0

In [117]:
out.loc["boat"].loc["Trump"].value_counts()[0]

6

In [118]:
out.size

70

In [49]:
out = lab.total_by_month(df)

In [50]:
out["Total"]["May"].idxmax()

('Smith', 'book')

In [51]:
out.shape[1]

5

# A distribution of Skittles

[Skittles](https://en.wikipedia.org/wiki/Skittles_(confectionery)) are made in two locations in the United States: Yorkville, Illinois and Waco, Texas. In these factories, Skittles of different colors are made separately by different machines and combined/packaged into bags for sale. The tab-separated file `skittles.tsv` contains the contents of 468 bags of Skittles.

Most people have preferences for their favorite flavor and there is a surprising amount of variation among the distribution of flavors in each bag.

Look at the variation by bag in the dataset below:

In [133]:
skittles_fp = os.path.join('data', 'skittles.tsv')
skittles = pd.read_csv(skittles_fp, sep='\t')

### Differences between Yorkville and Waco

**Question 4**

First, you will investigate if the machine that mixes together the Skittles of different colors might favor one color over another. Use a permutation test to assess whether, on average, bags made in Yorkville have the same number of orange skittles as bags made in Waco. Do this by following the outline below:
1. Create a function `diff_of_means` that takes in a dataframe of counts of skittles (like `skittles`) and their origin and returns the *absolute* difference of means between the number of orange Skittles per bag from Yorkville and Waco.
2. Create a function `simulate_null` that takes in a dataframe of counts of skittles (like `skittles`) and their origin, and returns one instance of the test-statistic under the null hypothesis.
3. Create a function `pval_orange` that takes in a dataframe of counts of skittles (like `skittles`) and their origin, and calculates the p-value for the permutation test using `1000` trials.

Plot the observed statistic, along with the histogram for the simulated distribution, to check your work.

In [52]:
skittles_fp = os.path.join('data', 'skittles.tsv')
skittles = pd.read_csv(skittles_fp, sep='\t')
# sr = skittles.groupby('Factory')['orange'].mean()
# abs(np.diff(sr.values)[0])
# abs(skittles.groupby('Factory')['orange'].mean().diff().iloc[-1])

In [53]:
out = lab.diff_of_means(skittles)
out

0.6004398826979482

In [54]:
0 <= out

True

In [55]:
skittles_fp = os.path.join('data', 'skittles.tsv')
skittles = pd.read_csv(skittles_fp, sep='\t')
out = lab.simulate_null(skittles)
out

0.6004398826979482

In [56]:
# sample = (skittles['orange'].sample(replace=False, frac=1).reset_index('Factory'))
# abs_diff = abs(sample.groupby('Factory')['orange'].mean().diff().iloc[-1])
# abs_diff

In [57]:
isinstance(out, float)

True

In [58]:
0 <= out <= 1.0

True

In [59]:
pval = lab.pval_orange(skittles)
pval

0.043

In [60]:
isinstance(pval, float)

True

In [61]:
0 <= pval <= 0.1

True

**Question 5**

Use your work from above to decide which colors tend to differ the most between the two locations on average. Create a function `ordered_colors` that returns your answer as an ordered list from "most different" to "least different" between the two locations. Your list should be a *hard-coded* list, where each element has the form `(color, p-value)`.

Even though there is randomness in the color composition in each bag, this list gives the likelihood that the machines have a systematic, meaningful, difference in how they blend the colors in each bag.

In [62]:
skittles_fp = os.path.join('data', 'skittles.tsv')
skittles = pd.read_csv(skittles_fp, sep='\t')
skittles.columns

Index(['red', 'orange', 'yellow', 'green', 'purple', 'Factory'], dtype='object')

In [63]:
orange_pval = lab.pval_orange(skittles)
orange_pval

0.046

In [64]:
red_pval = lab.pval_orange(skittles)
red_pval

0.228

In [65]:
yellow_pval = lab.pval_orange(skittles)
yellow_pval

0.0

In [68]:
green_pval = lab.pval_orange(skittles)
green_pval

0.477

In [67]:
purple_pval = lab.pval_orange(skittles)
purple_pval

0.973

In [167]:
out = lab.ordered_colors()
out

(('purple', 0.468),
 ('green', 0.227),
 ('yellow', 0.0),
 ('red', 0.1),
 ('orange', 0.019))

In [70]:
out = lab.ordered_colors()
out

(('purple', 0.973),
 ('green', 0.477),
 ('red', 0.228),
 ('orange', 0.046),
 ('yellow', 0.0))

In [168]:
len(out) == 5

True

In [169]:
colors = {'green', 'orange', 'purple', 'red', 'yellow'}

In [170]:
set([x[0] for x in out]) == colors

True

In [171]:
all([isinstance(x[1], float) for x in out])

True

**Question 6**

Now, suppose you would like to assess whether the two locations make similar amounts of each color overall. That is:
* Combine and count up all the Skittles of each color that were made in Yorkville 
* Combine and count up all the Skittles of each color that were made in Waco

Are these distributions of colors similar? Is the variation among the bags due to each factory making different amounts of each color?

Use a permutation test to assess whether the distribution of colors of Yorkville Skittles is statistically significantly different than those made in Waco. Set a significance level of 0.01 and determine whether you can reject a null hypothesis that answer the question above using a permutation test with 1000 trials. For your test statistic, use the total-variation-distance (TVD).

Create a function `same_color_distribution` of zero variables that outputs a hard-coded tuple with the p-value and whether you 'Fail to Reject' or 'Reject' the null hypothesis.

For this question, the following references may be useful:
* For TVD reference, see [DSC 10](https://www.inferentialthinking.com/chapters/11/2/Multiple_Categories.html) and [Lecture 04](https://github.com/ucsd-ets/dsc80-wi20/tree/master/lectures/04)
* For permutation test reference, see [DSC 10](https://www.inferentialthinking.com/chapters/12/Comparing_Two_Samples.html) and [Lecture 07](https://github.com/ucsd-ets/dsc80-wi20/tree/master/lectures/07)


In [259]:
skittles_fp = os.path.join('data', 'skittles.tsv')
skittles = pd.read_csv(skittles_fp, sep='\t')

skittles = skittles.groupby('Factory').sum()
distr = skittles.apply(lambda x: x / x.sum())
observe = (distr.diff(axis=1).iloc[-1].abs().sum()) / 2

In [260]:
skittles_fp = os.path.join('data', 'skittles.tsv')
skittles = pd.read_csv(skittles_fp, sep='\t')

test_distr = []
for _ in range(1000):
    shuffle_factory = (skittles['Factory'].sample(replace=False, frac=1).reset_index(drop= True))
    shuffled = (skittles.assign(**{'Shuffled factory': shuffle_factory}))
    group_sum = shuffled.groupby('Shuffled factory').sum()
    distr = group_sum.apply(lambda x: x / x.sum())
    pval = (distr.diff(axis=1).iloc[-1].abs().sum()) / 2
    test_distr.append(pval)

In [261]:
p_value = np.count_nonzero(test_distr >= observe) / 1000
p_value

0.005

In [253]:
out = lab.same_color_distribution()

In [254]:
isinstance(out[0], float)

True

In [255]:
out[1] in ['Fail to Reject', 'Reject']

True

### Hypothesis vs Permutation Testing

**Question 7**

In each of the following scenarios, decide  whether  a  permutation test is appropriate to determine if there is a  significant difference between the quantities described. If a permutation test is appropriate, mark 'P'. Otherwise, mark 'H'.

Record your answers in the function `perm_vs_hyp` that outputs a list of length 5, containing the values 'H' and 'P"

1. Compare the DSC 80 pass rate between second years and third years who take the class.
    - Permutation or Hypothesis test?
2. Compare the proportion of Data Science majors who have completed DSC 80 and the proportion of Data Science minors who have completed DSC 80.
    - Permutation or Hypothesis test?
3. Compare the proportion of students who have iPhones to the proportion of students who have Android phones.
    - Permutation or Hypothesis test?
4. Out of the DSC 80 students, the professor asks students whether they prefer DSC 10 or DSC 20.  Compare the proportion of students who prefer DSC 10 to the proportion who prefer DSC 20.
    - Permutation or Hypothesis test?
5. Compare the attendance rate of classes that use iClickers vs classes that do not use iClickers.
    - Permutation or Hypothesis test?

In [268]:
out = lab.perm_vs_hyp()

In [269]:
ans = ['P', 'H']

In [270]:
len(out) == 5

True

In [271]:
set(out) <= set(ans)

True

# Types of Missingness

### Missing by Design (MD)
- The missing field is deliberately missing. The missing field is deliberately not collected or set to null (hence, "missing by design")
- The missingness can be exactly predicted when a column will be null, with only knowledge of the other columns using a function of the rows of the dataset

### Missing Completely at Random (MCAR)
- The missingness of missing value isn't related to the actual, unreported value itself, nor the values in any other fields. The missingness is not systematic.
- The missingness is unconditionally uniform across rows. MCAR doesn't bias the observed data.
- There is no relationship between the missing data and the any of the other data, observed or missing.

### Missing at Random (MAR)
- The missingness of the missing value has nothing to do with the value itself, but may be related to another field.
- The missingness is uniform across rows, perhaps conditional on another column. MAR biases the observed data, but is fixable.
- There is a systematic relationship between the missing values and the observed data (but not the missing values themselves).
- Difference between MD and MAR: If you can *exactly/always* determine missingness on other columns, the missingness is MD. If there is just some sort of systematic relationship between the missing columns/values and other columns/values that may help us predict missingness, the missingness is MAR.

### Non-Ignorable (NI)
- The missingness of the missing value is related to the actual, unreported value.
- NI biases the observed data in unobservable ways.
- There is relationship between the propensity of a value to be missing and its value.

*Note:* In class, we sometimes refer to non-ignorable missingness as "not missing at random (NMAR)"

# Missingness Mechanisms

For each of the following scenarios, choose the most likely mechanism for the missing data in the dataset described.

### After-purchase surveys

**Question 8**

You run a small e-commerce website and send surveys out to customers after they purchase an item from your store. The survey asks whether the customer is satisfied with their purchase ("Yes" or "No"). Below, you are presented with possible datasets, each of which contains a column `satisfied` as described above, as well as a `customer_id` number corresponding to the order. The column `satisfied` is missing data. 

For each of the following datasets, label the column `satisfied` as being `MD`, `MCAR`, `MAR`, `NI`.

1. The dataset consists only of the columns `customer_id` and `satisfied`.
2. The dataset contains the `customer_id` of every customer with an account, even if they didn't make a purchase. Also, in this case, you notice everyone who was sent a survey filled it out.
3. The dataset contains a column specifying if the user later returned the item.
4. The dataset contains a column with the serial number for the item purchased.
5. The dataset contains a column with the price of the item purchased.

Record your answers in the function `after_purchase` that outputs a list of length 5, containing the values `MD`, `MCAR`, `MAR`, `NI`.

In [ ]:
#1 NI

In [ ]:
#2 MD

In [ ]:
#3 MAR

In [ ]:
#4 MCAR

In [ ]:
#5 MAR

In [277]:
out = lab.after_purchase()

In [278]:
ans = ['MD', 'MCAR', 'MAR', 'NI']

In [279]:
len(out) == 5

True

In [280]:
set(out) <= set(ans)

True

### Miscellaneous missingness questions

**Question 9**

In each of the following scenarios, choose the best answer. Return your answers in a function `multiple_choice`.

1. UCSD has recently adopted GrubHub as the food pre-ordering app for campus restaurants, so you can order your food ahead of time and stop by before your next class. In a table of GrubHub app orders, which contains information such as `restaurant`, `name`, `items`, and `total`, the column `delivery_address` is often missing for UCSD students. The missingness mechanism of these columns is likely:
    - Is the exam grade column `MD`, `MCAR`, `MAR`, `NI`?
1. In a database of student records that records student profile data, such as `name`, `home_address`, `ethnicity`, etc., sometimes the Middle Name column is missing. This column is most likely:
    - `MD`, `MCAR`, `MAR`, `NI`
1. The UCSD Club Basketball team creates a signup sheet for potential new members. The sheet contains the columns: `full_name`, `year`, `email`, `favorite_sports`, `number_of_sports_played`, `sports_previously_played`. The team president notices that many students left the `sports_previously_played` blank. The missingness mechanism of this column is likely:
    - `MD`, `MCAR`, `MAR`, `NI`
1. Associated Students sends out a survey to all students about their 2019 Sun God Experience, with all questions being optional. They notice that many students left the "Were you satisfied with Sun God 2019?". This missingness is most likely:
    - `MD`, `MCAR`, `MAR`, `NI`
1. UC San Diego is implementing two-step login through DUO on October 16th. On October 1st, an administrator creates a table of randomized student codes for all students (not their PID) and the phone numbers associated with their DUO account (two columns total). The administrator notices that there is a lot of missingness in the phone numbers column. This column is most likely:
    - `MD`, `MCAR`, `MAR`, `NI`

In [ ]:
#1 MCAR //MAR

In [ ]:
#2 NI

In [ ]:
#3 MD

In [ ]:
#4 NI

In [ ]:
#5 MAR //MCAR

## Congratulations! You're done!

* Submit the lab on Gradescope